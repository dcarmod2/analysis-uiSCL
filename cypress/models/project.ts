import Bundle from './bundle'
import Model from './model'
import Modification from './modification'

type ModificationProps = {
  data?: Record<string, unknown>
  onCreate?: () => void
  name?: string
  type: Cypress.ModificationType
}

export default class Project extends Model {
  bundle: Bundle
  modificationNames: Set<string> = new Set()

  constructor(parentKey: string, name: string, bundle: Bundle) {
    super(parentKey, name, 'project')
    this.bundle = bundle
  }

  _delete() {
    cy.findByLabelText('Edit project settings').click()
    cy.navComplete()
    cy.findByText(/Delete project/i).click()
    cy.findByRole('button', {name: /Confirm: Delete project/i}).click()
  }

  deleteModification(modificationName: string) {
    this.navTo()
    // Create if it does not exist
    cy.findAllByRole('button').then((buttons) => {
      const pb = buttons.filter((_, el) => el.textContent === modificationName)
      if (pb.length !== 0) {
        cy.wrap(pb.first()).click()
        cy.findByRole('button', {name: /Delete modification/}).click()
        cy.findByRole('button', {name: /Confirm: Delete modification/}).click()
        cy.navComplete()
      }
    })
  }

  // Create a project if it does not exist.
  findOrCreate() {
    cy.navTo('projects')
    cy.get('button').then((buttons) => {
      const pb = buttons.filter((_, el) => el.textContent === this.name)
      if (pb.length === 0) {
        cy.findByText(/Create new Project/i).click()
        cy.findByLabelText(/Project name/).type(this.name)
        cy.findByLabelText(/Associated network bundle/i)
          .click({force: true})
          .type(this.bundle.name + '{enter}')
        cy.findByText(/^Create$/).click()
      } else {
        cy.wrap(pb.first()).click()
      }

      // Store the project id
      cy.location('pathname')
        .should('match', /regions\/\w{24}\/projects\/\w{24}\/modifications$/)
        .then((path) => {
          this.path = path
        })
    })
  }

  getModification({
    data,
    onCreate,
    name,
    type
  }: ModificationProps): Modification {
    name ??= type + this.modificationNames.size
    if (this.modificationNames.has(name)) {
      throw new Error('Modification name must be unique per project')
    } else {
      this.modificationNames.add(name)
    }
    const modification = new Modification(this.key, name, type, onCreate)

    before(`getModification(${modification.name})`, () => {
      this.navTo()
      modification.initialize()

      // On each test run ensure it starts with the same data.
      if (data) {
        modification.navTo()
        cy.editModificationJSON(data)
      }
    })

    return modification
  }

  // Helper for setting up a scenarios in this project.
  getScenarios(scenarios: string[]) {
    before('getScenarios', () => {
      this.navTo()
      cy.findByRole('tab', {name: 'Scenarios'}).click()
      scenarios.forEach((scenarioName) => {
        cy.get('#scenarios').then((el) => {
          // create named scenario if it doesn't already exist
          if (!el.text().includes(scenarioName)) {
            cy.findByRole('button', {name: 'Create a scenario'}).click()
            cy.loadingComplete()
            // TODO there has GOT to be a better way...
            cy.get('#scenarios').findAllByRole('group').last().click()

            cy.focused()
              .clear()
              .type(scenarioName + '{enter}')
          }
        })
      })
    })
  }

  navTo() {
    cy.navComplete()
    cy.location('pathname').then((path) => {
      if (path === this.path) {
        cy.findByRole('tab', {name: /Modifications/}).click()
      } else {
        cy.visit(this.path)
        cy.navComplete()
      }
    })
  }
}
