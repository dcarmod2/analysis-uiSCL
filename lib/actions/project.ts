import {stringify} from 'querystring'
import {createAction} from 'redux-actions'

import {API} from 'lib/constants'
import selectCurrentProject from 'lib/selectors/current-project'

import fetch from './fetch'
import getFeedsRoutesAndStops from './get-feeds-routes-and-stops'
import {
  saveToServer as saveModification,
  getForProject as getModificationsForProject
} from './modifications'

import {loadBundle} from './'

const PROJECT_URL = API.Project

// project stuff
const addProject = createAction('add project')
export const create = ({bundleId, name, regionId}) =>
  fetch({
    url: PROJECT_URL,
    options: {
      body: {
        bundleId,
        name,
        regionId,
        variants: ['Default']
      },
      method: 'post'
    },
    next: (response) => addProject(response.value)
  })

const deleteLocally = createAction('delete project')
export const deleteProject = (projectId) =>
  fetch({
    options: {
      method: 'delete'
    },
    url: `${PROJECT_URL}/${projectId}`,
    next: () => deleteLocally(projectId)
  })

export const loadProject = (_id) =>
  fetch({
    url: `${PROJECT_URL}/${_id}`,
    next: (r) => set(r.value)
  })

export const loadProjectAndModifications = (_id) => async (dispatch) => {
  const r1 = await Promise.all([
    dispatch(loadProject(_id)),
    dispatch(getModificationsForProject(_id))
  ])
  const project = r1[0]
  const modifications = r1[1]

  const r2 = await Promise.all([
    dispatch(loadBundle(project.bundleId)),
    dispatch(
      getFeedsRoutesAndStops({
        bundleId: project.bundleId,
        forceCompleteUpdate: true,
        modifications
      })
    )
  ])

  return {bundle: r2[0], feeds: r2[1], modifications, project}
}

export const saveToServer = (project) =>
  fetch({
    url: `${PROJECT_URL}/${project._id}`,
    options: {
      body: project,
      method: 'put'
    },
    next: (response) => set(response.value)
  })
export const set = createAction('set project')
export const setAll = createAction('set projects')
export const loadProjects = (query = {}) =>
  fetch({
    url: `${PROJECT_URL}?${stringify(query)}`,
    next: (response) => setAll(response.value)
  })

/**
 * Project Variants
 */
export const createVariant = (name) => (dispatch, getState) => {
  const project = selectCurrentProject(getState())
  const {variants} = project
  return dispatch(
    saveToServer({
      ...project,
      variants: [...variants, name]
    })
  )
}

/**
 * Create a new scenario with all of the same modifications,
 */
export const copyScenario = (index) => async (dispatch, getState) => {
  const state = getState()
  const project = selectCurrentProject(state)
  const {modifications} = state.project

  const newScenarioName = project.variants[index] + ' (copy)'
  const oldLength = project.variants.length

  await dispatch(
    saveToServer({
      ...project,
      variants: [...project.variants, newScenarioName]
    })
  )

  return Promise.all(
    modifications.map((m) => {
      const variants = m.variants
      // Could come from a bitset on the Java side so may be of varying length
      for (let i = 0; i < oldLength; i++) {
        if (variants[i] === undefined) variants[i] = false
      }

      return dispatch(
        saveModification({
          ...m,
          variants: [
            ...variants,
            variants[index] // add the new variant
          ]
        })
      )
    })
  )
}

export const deleteVariant = (index) => async (dispatch, getState) => {
  const state = getState()
  const project = selectCurrentProject(state)
  const {modifications} = state.project

  await dispatch(
    saveToServer({
      ...project,
      variants: project.variants.filter((_, i) => i !== index)
    })
  )

  return Promise.all(
    modifications.map((m) =>
      dispatch(
        saveModification({
          ...m,
          variants: m.variants.filter((_, i) => i !== index)
        })
      )
    )
  )
}

export const editVariantName = ({index, name}) => (dispatch, getState) => {
  const project = selectCurrentProject(getState())
  return dispatch(
    saveToServer({
      ...project,
      variants: project.variants.map((value, i) => (i === index ? name : value))
    })
  )
}
