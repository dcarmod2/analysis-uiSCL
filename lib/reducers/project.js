import sortBy from 'lodash/sortBy'

function createInitialState() {
  return {
    modifications: [],
    feeds: [],
    projects: []
  }
}

const sortByTypeName = (a) => sortBy(a, ['type', 'name'])
const sortByCreatedAt = (a) => sortBy(a, (i) => -i.createdAt)

export const reducers = {
  /** add a newly created project to the state */
  'add project'(state, action) {
    return {
      ...state,
      projects: [action.payload, ...state.projects]
    }
  },
  'delete modification'(state, action) {
    const modificationId = action.payload
    return {
      ...state,
      modifications: state.modifications.filter((m) => m._id !== modificationId)
    }
  },
  'delete project'(state, action) {
    return {
      ...state,
      projects: state.projects.filter((p) => p._id !== action.payload)
    }
  },
  'log out'() {
    return createInitialState()
  },
  'set feeds'(state, action) {
    return {
      ...state,
      feeds: action.payload || []
    }
  },
  'create modification'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName([action.payload, ...state.modifications])
    }
  },
  'create modifications'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName([...state.modifications, ...action.payload])
    }
  },
  'set modification'(state, action) {
    const modification = action.payload
    const modifications = state.modifications.filter(
      (m) => m._id !== modification._id
    )
    return {
      ...state,
      modifications: sortByTypeName([...modifications, modification])
    }
  },
  'set modifications'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName(action.payload)
    }
  },
  'update modification'(state, action) {
    const {_id, ...newProps} = action.payload
    return {
      ...state,
      modifications: state.modifications.map((m) =>
        m._id === _id ? {...m, ...newProps} : m
      )
    }
  },
  'set project'(state, action) {
    const {projects} = state
    const currentProject = {...action.payload}
    return {
      ...state,
      projects: sortByCreatedAt([
        ...projects.filter((p) => p._id !== currentProject._id),
        currentProject
      ])
    }
  },
  'set projects'(state, action) {
    return {
      ...state,
      projects: action.payload
    }
  },
  'clear current region'() {
    return createInitialState()
  }
}

export const initialState = createInitialState()
