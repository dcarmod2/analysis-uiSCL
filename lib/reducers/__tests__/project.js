//
import {handleActions} from 'redux-actions'

import * as project from '../project'

describe('reducers > project', () => {
  const reducer = handleActions(project.reducers, project.initialState)
  const mockIdObject = {_id: '1'}
  const mockModification = {_id: '1', type: 'adjust-dwell-time'}

  // Default State Test
  it('should handle default state', () => {
    expect(reducer(undefined, {type: 'blah', payload: {}})).toMatchSnapshot()
  })

  // Specific Handler Tests
  it('should handle add bundle', () => {
    const action = {type: 'add bundle', payload: mockIdObject}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle add project', () => {
    const action = {type: 'add project', payload: mockIdObject}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle delete bundle', () => {
    const action = {type: 'delete bundle', payload: '1'}
    const beginState = {bundles: [mockIdObject]}
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle delete modification', () => {
    const action = {type: 'delete modification', payload: '1'}
    const beginState = {modifications: [mockModification]}
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle delete project', () => {
    const action = {type: 'delete project', payload: '1'}
    const beginState = {projects: [mockIdObject]}
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle log out', () => {
    expect(reducer(undefined, {type: 'log out', payload: {}})).toMatchSnapshot()
  })

  it('should handle set bundle', () => {
    const action = {type: 'set bundle', payload: '1'}
    const beginState = {
      bundles: [mockIdObject],
      currentProject: {}
    }
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle set bundles', () => {
    const action = {type: 'set bundles', payload: [mockIdObject]}
    const beginState = {
      currentProject: {
        bundleId: '1'
      }
    }
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle set feeds', () => {
    const action = {type: 'set feeds', payload: [mockIdObject]}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set modification', () => {
    const action = {type: 'set modification', payload: mockModification}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set modifications', () => {
    const action = {type: 'set modifications', payload: [mockModification]}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set project', () => {
    const mockProject = {_id: '1', bundleId: '1'}
    const action = {type: 'set project', payload: mockProject}
    const beginState = {bundles: [mockIdObject], projects: []}
    expect(reducer(beginState, action)).toMatchSnapshot()
  })

  it('should handle set projects', () => {
    const action = {type: 'set projects', payload: [mockIdObject]}
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle show variant', () => {
    const action = {type: 'show variant', payload: 1}
    const beginState = {
      modifications: [
        {
          _id: '1',
          variants: [],
          type: 'adjust-dwell-time'
        }
      ]
    }
    expect(reducer(beginState, action)).toMatchSnapshot()
  })
})
