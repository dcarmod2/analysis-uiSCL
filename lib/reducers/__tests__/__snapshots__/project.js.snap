// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers > project should handle add bundle 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [],
  "projects": Array [],
}
`;

exports[`reducers > project should handle add project 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [],
  "projects": Array [
    Object {
      "_id": "1",
    },
  ],
}
`;

exports[`reducers > project should handle default state 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [],
  "projects": Array [],
}
`;

exports[`reducers > project should handle delete bundle 1`] = `
Object {
  "bundles": Array [
    Object {
      "_id": "1",
    },
  ],
}
`;

exports[`reducers > project should handle delete modification 1`] = `
Object {
  "modifications": Array [],
}
`;

exports[`reducers > project should handle delete project 1`] = `
Object {
  "projects": Array [],
}
`;

exports[`reducers > project should handle log out 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [],
  "projects": Array [],
}
`;

exports[`reducers > project should handle set bundle 1`] = `
Object {
  "bundles": Array [
    Object {
      "_id": "1",
    },
  ],
  "currentProject": Object {},
}
`;

exports[`reducers > project should handle set bundles 1`] = `
Object {
  "currentProject": Object {
    "bundleId": "1",
  },
}
`;

exports[`reducers > project should handle set feeds 1`] = `
Object {
  "feeds": Array [
    Object {
      "_id": "1",
    },
  ],
  "modifications": Array [],
  "projects": Array [],
}
`;

exports[`reducers > project should handle set modification 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [
    Object {
      "_id": "1",
      "type": "adjust-dwell-time",
    },
  ],
  "projects": Array [],
}
`;

exports[`reducers > project should handle set modifications 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [
    Object {
      "_id": "1",
      "type": "adjust-dwell-time",
    },
  ],
  "projects": Array [],
}
`;

exports[`reducers > project should handle set project 1`] = `
Object {
  "bundles": Array [
    Object {
      "_id": "1",
    },
  ],
  "projects": Array [
    Object {
      "_id": "1",
      "bundleId": "1",
    },
  ],
}
`;

exports[`reducers > project should handle set projects 1`] = `
Object {
  "feeds": Array [],
  "modifications": Array [],
  "projects": Array [
    Object {
      "_id": "1",
    },
  ],
}
`;

exports[`reducers > project should handle show variant 1`] = `
Object {
  "modifications": Array [
    Object {
      "_id": "1",
      "type": "adjust-dwell-time",
      "variants": Array [],
    },
  ],
}
`;
